staking/staking_hub.go

```diff
@@ -0,0 +1,394 @@
+	byteValue, _ := ioutil.ReadAll(mapFile)
+
+	var stakingMap []StakerMapping
+
+	// Unmarshal the json
+	if err = json.Unmarshal(byteValue, &stakingMap); err != nil {
+		return fmt.Errorf("unable to unmarshal staking map")
+	}
+
+	// Close the file
+	if err = mapFile.Close(); err != nil {
+		return fmt.Errorf("unable to close staking file")
+	}
+
+	// Update the staking map using the read data
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	for _, val := range stakingMap {
+		sh.StakingMap[val.Address] = val.Stake
+	}
+
+	return nil
+}
+
+// marshalJSON generates the json object for staker mappings
+func (sh *StakingHub) marshalJSON(mappings []StakerMapping) (io.Reader, error) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	b, err := json.MarshalIndent(mappings, "", "\t")
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(b), nil
+}
+
+// isStaker is a helper method to check whether or not an address has a staked balance
+func (sh *StakingHub) isStaker(address types.Address) bool {
+	if _, ok := sh.StakingMap[address]; ok {
+		return true
+	}
+
+	return false
+}
+
+// IncreaseStake increases the account's staked balance, or sets it if the account wasn't
+// in the StakingMap
+func (sh *StakingHub) IncreaseStake(address types.Address, stakeBalance *big.Int) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if !sh.isStaker(address) {
+		sh.StakingMap[address] = stakeBalance
+	} else {
+		sh.StakingMap[address] = big.NewInt(0).Add(sh.StakingMap[address], stakeBalance)
+	}
+
+	sh.log(
+		fmt.Sprintf("Stake increase:\t%s %s", address.String(), stakeBalance.String()),
+		logInfo,
+	)
+}
+
+// DecreaseStake decreases the account's staked balance if the account is present
+func (sh *StakingHub) DecreaseStake(address types.Address, unstakeBalance *big.Int) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) && sh.StakingMap[address].Cmp(unstakeBalance) >= 0 {
+		sh.StakingMap[address] = big.NewInt(0).Sub(sh.StakingMap[address], unstakeBalance)
+
+		sh.log(
+			fmt.Sprintf("Stake decrease:\t%s %s", address.String(), unstakeBalance.String()),
+			logInfo,
+		)
+	}
+}
+
+// ResetStake resets the account's staked balance
+func (sh *StakingHub) ResetStake(address types.Address) {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		delete(sh.StakingMap, address)
+	}
+
+	sh.log(
+		fmt.Sprintf("Stake reset:\t%s", address.String()),
+		logInfo,
+	)
+}
+
+// GetStakedBalance returns an accounts staked balance if it is a staker.
+// Returns 0 if the address is not a staker
+func (sh *StakingHub) GetStakedBalance(address types.Address) *big.Int {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	if sh.isStaker(address) {
+		return sh.StakingMap[address]
+	}
+
+	return big.NewInt(0)
+}
+
+// GetStakerAddresses returns a list of all addresses that have a stake > 0
+func (sh *StakingHub) GetStakerAddresses() []types.Address {
```

Comment:
You can remove this method since it's unused.

Reply:
Removing now 👍